config/store.go

```diff
@@ -56,11 +67,273 @@ type Store interface {
 		return NewDatabaseStore(dsn)
 	}
 
 	return NewFileStore(dsn, watch)
 }
+
+func NewTestMemoryStore() Store {
+	memoryStore, err := NewMemoryStore()
+	if err != nil {
+		panic("failed to initialize memory store: " + err.Error())
+	}
+
+	configStore, err := NewStoreFromBacking(memoryStore)
+	if err != nil {
+		panic("failed to initialize config store: " + err.Error())
+	}
+
+	return configStore
+}
+
+type storeImpl struct {
+	emitter
+	backingStore BackingStore
+
+	configLock  sync.RWMutex
+	config      *model.Config
+	configNoEnv *model.Config
+
+	persistFeatureFlags bool
+}
+
+// Get fetches the current, cached configuration.
+func (s *storeImpl) Get() *model.Config {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return s.config
+}
+
+// Get fetches the current, cached configuration without environment variable overrides.
+func (s *storeImpl) GetNoEnv() *model.Config {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return s.configNoEnv
+}
+
+// GetEnvironmentOverrides fetches the configuration fields overridden by environment variables.
+func (s *storeImpl) GetEnvironmentOverrides() map[string]interface{} {
+	return generateEnviromentMap(GetEnviroment())
+}
+
+// RemoveEnvironmentOverrides returns a new config without the environment
+// overrides
+func (s *storeImpl) RemoveEnvironmentOverrides(cfg *model.Config) *model.Config {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return removeEnvOverrides(cfg, s.configNoEnv, s.GetEnvironmentOverrides())
+}
+
+// PersistFeatures sets if the store should persist feature flags.
+func (s *storeImpl) PersistFeatures(persist bool) {
+	s.configLock.Lock()
+	defer s.configLock.Unlock()
+	s.persistFeatureFlags = persist
+}
+
+// Set replaces the current configuration in its entirety and updates the backing store.
+func (s *storeImpl) Set(newCfg *model.Config) (*model.Config, error) {
+	s.configLock.Lock()
+	var unlockOnce sync.Once
+	defer unlockOnce.Do(s.configLock.Unlock)
+
+	oldCfg := s.config.Clone()
+
+	// Really just for some tests we need to set defaults here
+	newCfg.SetDefaults()
+
+	// Sometimes the config is received with "fake" data in sensitive fields. Apply the real
+	// data from the existing config as necessary.
+	desanitize(oldCfg, newCfg)
+
+	if err := newCfg.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "new configuration is invalid")
+	}
+
+	newCfg = removeEnvOverrides(newCfg, s.configNoEnv, s.GetEnvironmentOverrides())
+
+	// Don't persist feature flags unless we are on MM cloud
+	// MM cloud uses config in the DB as a cache of the feature flag
+	// settings in case the managment system is down when a pod starts.
+	if !s.persistFeatureFlags {
+		newCfg.FeatureFlags = nil
+	}
+
+	if err := s.backingStore.Set(newCfg); err != nil {
+		return nil, errors.Wrap(err, "failed to persist")
+	}
+
+	if err := s.loadLockedWithOld(oldCfg, &unlockOnce); err != nil {
+		return nil, err
```

Comment:
nit: a few errors around here that could be wrapped if it is deemed helpful.

Reply:
Wrapped some errors.